server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java

```diff
@@ -72,9 +72,14 @@ public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float bo
             throw new IllegalStateException("This KnnScoreDocQuery was created by a different reader");
         }
         return new Weight(this) {
+            @Override
+            public int count(LeafReaderContext context) {
+                return segmentStarts[context.ord + 1] - segmentStarts[context.ord];
```

Comment:
We can optimize our `count` since `segmentStarts` indicate the number of documents within that segment. So, count can be constant time. This is a minor optimization, but prevents creating a scorer and iterating docs needlessly.

Reply:
I think it would be wise to make this change separately. it's a good one but it's a performance improvement rather than a bugfix.